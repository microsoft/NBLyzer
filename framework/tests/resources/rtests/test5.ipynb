{"cells":[{"cell_type":"code","execution_count":1,"metadata":{"_cell_guid":"f67b9393-8ea1-4e23-b856-2ce149cfe421","_execution_state":"idle","_uuid":"72334cb006d02a4bcfc2a2fe622524eba824c6f8","collapsed":true},"outputs":[],"source":["import pandas as pd\n","import numpy as np\n","import matplotlib.pyplot as plt\n","import matplotlib.image as mpimg\n","import seaborn as sns\n","%matplotlib inline\n","\n","np.random.seed(2)\n","\n","from sklearn.model_selection import train_test_split\n","from sklearn.metrics import confusion_matrix\n","import itertools\n","\n","from keras.utils.np_utils import to_categorical # convert to one-hot-encoding\n","from keras.models import Sequential\n","from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPool2D\n","from keras.optimizers import RMSprop\n","from keras.preprocessing.image import ImageDataGenerator\n","from keras.callbacks import ReduceLROnPlateau\n","\n","\n","sns.set(style='white', context='notebook', palette='deep')"]},{"cell_type":"code","execution_count":2,"metadata":{"_cell_guid":"5e51d00e-62fd-4141-bf73-50ac4f2da7d0","_execution_state":"idle","_uuid":"84bbd5ab8d7895bd430d5ecfe2f7ddf77baa7b74","collapsed":true},"outputs":[],"source":["# Load the data\n","train = pd.read_csv(\"../input/train.csv\")\n","test = pd.read_csv(\"../input/test.csv\")"]},{"cell_type":"code","execution_count":3,"metadata":{"_cell_guid":"86570a36-5c20-460a-9dfd-2070548532a7","_execution_state":"idle","_uuid":"1213b979d5ed3e0d13824d17d694c79d2ece92fa","collapsed":true},"outputs":[],"source":["Y_train = train[\"label\"]\n","\n","# Drop 'label' column\n","X_train = train.drop(labels = [\"label\"],axis = 1) \n","\n","# free some space\n","del train \n","\n","g = sns.countplot(Y_train)\n","\n","Y_train.value_counts()"]},{"cell_type":"code","execution_count":4,"metadata":{"_cell_guid":"ececaa00-2ae3-4d13-b631-438df085b030","_execution_state":"idle","_uuid":"cdf27c27e2a5b15e6d7bfc70de7a18c08f3feb7a","collapsed":true},"outputs":[],"source":["# Check the data\n","X_train.isnull().any().describe()"]},{"cell_type":"code","execution_count":5,"metadata":{"_cell_guid":"92572e80-8543-4e72-8767-5c9be8381b04","_execution_state":"idle","_uuid":"a0089bb7ec9aec76373db475399aea24699ae989","collapsed":true},"outputs":[],"source":["test.isnull().any().describe()"]},{"cell_type":"code","execution_count":6,"metadata":{"_cell_guid":"cdc4340b-6e24-4e12-be99-ac806098ff17","_execution_state":"idle","_uuid":"b5d4f8fcf2a967e2c7d57daedf95aa8c5ab7f8cb","collapsed":true},"outputs":[],"source":["# Normalize the data\n","X_train = X_train / 255.0\n","test = test / 255.0"]},{"cell_type":"code","execution_count":7,"metadata":{"_cell_guid":"34b6a5f7-8fd2-4387-8ef4-c9dc19584fed","_execution_state":"idle","_uuid":"f0a6ad80dab8e0f2c2e46165ccd9cd82dd162bc3","collapsed":true},"outputs":[],"source":["# Reshape image in 3 dimensions (height = 28px, width = 28px , canal = 1)\n","X_train = X_train.values.reshape(-1,28,28,1)\n","test = test.values.reshape(-1,28,28,1)"]},{"cell_type":"code","execution_count":8,"metadata":{"_cell_guid":"4b7f3e78-44dc-4561-b1f0-9429ee024cf4","_execution_state":"idle","_uuid":"cabefd1478d5c1bdfe57fd6a34395340916a854c","collapsed":true},"outputs":[],"source":["# Encode labels to one hot vectors (ex : 2 -> [0,0,1,0,0,0,0,0,0,0])\n","Y_train = to_categorical(Y_train, num_classes = 10)"]},{"cell_type":"code","execution_count":9,"metadata":{"_cell_guid":"3a698301-9759-4279-ae48-fd980f89ea53","_execution_state":"idle","_uuid":"6e51c925c6e0f1b936679c9649fef345c853555f","collapsed":true},"outputs":[],"source":["# Set the random seed\n","random_seed = 2"]},{"cell_type":"code","execution_count":10,"metadata":{"_cell_guid":"dcd25ebb-d845-4d32-9867-082e352b1396","_execution_state":"idle","_uuid":"b779ac76d8317647db92d5a88b4098d212d72884","collapsed":true},"outputs":[],"source":["# Split the train and the validation set for the fitting\n","X_train, X_val, Y_train, Y_val = train_test_split(X_train, Y_train, test_size = 0.1, random_state=random_seed)"]},{"cell_type":"code","execution_count":11,"metadata":{"_cell_guid":"5f76131b-4ba0-45f1-a98c-bd4e7d561793","_execution_state":"idle","_uuid":"e0dae8943d3d35f075dba3d7ba31bde1d4bf2ff4","collapsed":true},"outputs":[],"source":["# Some examples\n","g = plt.imshow(X_train[0][:,:,0])"]},{"cell_type":"code","execution_count":12,"metadata":{"_cell_guid":"1e0f3f88-2ad7-459e-8e02-aecc5f3511ae","_execution_state":"idle","_uuid":"f7991ef6871a26f9fa57acdcd460a69bab53e804","collapsed":true},"outputs":[],"source":["# Set the CNN model \n","# my CNN architechture is In -> [[Conv2D->relu]*2 -> MaxPool2D -> Dropout]*2 -> Flatten -> Dense -> Dropout -> Out\n","\n","model = Sequential()\n","\n","model.add(Conv2D(filters = 32, kernel_size = (5,5),padding = 'Same', \n","                 activation ='relu', input_shape = (28,28,1)))\n","model.add(Conv2D(filters = 32, kernel_size = (5,5),padding = 'Same', \n","                 activation ='relu'))\n","model.add(MaxPool2D(pool_size=(2,2)))\n","model.add(Dropout(0.25))\n","\n","\n","model.add(Conv2D(filters = 64, kernel_size = (3,3),padding = 'Same', \n","                 activation ='relu'))\n","model.add(Conv2D(filters = 64, kernel_size = (3,3),padding = 'Same', \n","                 activation ='relu'))\n","model.add(MaxPool2D(pool_size=(2,2), strides=(2,2)))\n","model.add(Dropout(0.25))\n","\n","\n","model.add(Flatten())\n","model.add(Dense(256, activation = \"relu\"))\n","model.add(Dropout(0.5))\n","model.add(Dense(10, activation = \"softmax\"))"]},{"cell_type":"code","execution_count":13,"metadata":{"_cell_guid":"a4c55409-6a65-400a-b5e8-a1dc535429c0","_execution_state":"idle","_uuid":"420c704367b397b8255fefe9d882b35ac8929b95","collapsed":true},"outputs":[],"source":["# Define the optimizer\n","optimizer = RMSprop(lr=0.001, rho=0.9, epsilon=1e-08, decay=0.0)"]},{"cell_type":"code","execution_count":14,"metadata":{"_cell_guid":"0bd89b79-1a74-40da-af9a-7a76a8b96ff2","_execution_state":"idle","_uuid":"b4c0288622227a3cd05479aa765e324dbb852f34","collapsed":true},"outputs":[],"source":["# Compile the model\n","model.compile(optimizer = optimizer , loss = \"categorical_crossentropy\", metrics=[\"accuracy\"])"]},{"cell_type":"code","execution_count":15,"metadata":{"_cell_guid":"b5987a18-6bbe-42a2-9d31-333ebc4f7af1","_execution_state":"idle","_uuid":"c4a5b4e462ec5362c47eef4fcc7956fd4e203307","collapsed":true},"outputs":[],"source":["# Set a learning rate annealer\n","learning_rate_reduction = ReduceLROnPlateau(monitor='val_acc', \n","                                            patience=3, \n","                                            verbose=1, \n","                                            factor=0.5, \n","                                            min_lr=0.00001)"]},{"cell_type":"code","execution_count":16,"metadata":{"_cell_guid":"970db455-b393-4b25-806d-92c6766c12c0","_execution_state":"idle","_uuid":"26b0647c46efdb6b1096cf7335a7bf2a3417543a","collapsed":true},"outputs":[],"source":["epochs = 1 # Turn epochs to 30 to get 0.9967 accuracy\n","batch_size = 86"]},{"cell_type":"code","execution_count":18,"metadata":{"_cell_guid":"b342befe-1a6f-44bf-8dab-28033a729122","_execution_state":"idle","_uuid":"21d6192c87d92d497c797656474bccd9cefc5647","collapsed":true},"outputs":[],"source":["# With data augmentation to prevent overfitting (accuracy 0.99286)\n","\n","datagen = ImageDataGenerator(\n","        featurewise_center=False,  # set input mean to 0 over the dataset\n","        samplewise_center=False,  # set each sample mean to 0\n","        featurewise_std_normalization=False,  # divide inputs by std of the dataset\n","        samplewise_std_normalization=False,  # divide each input by its std\n","        zca_whitening=False,  # apply ZCA whitening\n","        rotation_range=10,  # randomly rotate images in the range (degrees, 0 to 180)\n","        zoom_range = 0.1, # Randomly zoom image \n","        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)\n","        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)\n","        horizontal_flip=False,  # randomly flip images\n","        vertical_flip=False)  # randomly flip images\n","\n","\n","datagen.fit(X_train)"]},{"cell_type":"code","execution_count":19,"metadata":{"_cell_guid":"b453af8d-9736-43e3-b486-7a1cd7dd8909","_execution_state":"idle","_uuid":"cf36b3d029f95b553be02d612e097a9769ee8252","collapsed":true},"outputs":[],"source":["# Fit the model\n","history = model.fit_generator(datagen.flow(X_train,Y_train, batch_size=batch_size),\n","                              epochs = epochs, validation_data = (X_val,Y_val),\n","                              verbose = 2, steps_per_epoch=X_train.shape[0] // batch_size\n","                              , callbacks=[learning_rate_reduction])"]},{"cell_type":"code","execution_count":20,"metadata":{"_cell_guid":"eb4b1b73-cbd4-40e8-9790-066fcef4c4c0","_execution_state":"idle","_uuid":"3a831860dd5bb65c8ead1ddcf4ae18ae20dd7f3e","collapsed":true},"outputs":[],"source":["# Plot the loss and accuracy curves for training and validation \n","fig, ax = plt.subplots(2,1)\n","ax[0].plot(history.history['loss'], color='b', label=\"Training loss\")\n","ax[0].plot(history.history['val_loss'], color='r', label=\"validation loss\",axes =ax[0])\n","legend = ax[0].legend(loc='best', shadow=True)\n","\n","ax[1].plot(history.history['acc'], color='b', label=\"Training accuracy\")\n","ax[1].plot(history.history['val_acc'], color='r',label=\"Validation accuracy\")\n","legend = ax[1].legend(loc='best', shadow=True)"]},{"cell_type":"code","execution_count":24,"metadata":{"_cell_guid":"11361e73-8250-4bf5-a353-b0f8ea83e659","_execution_state":"idle","_uuid":"16e161179bf1b51ba66c39b2cead883f1db3a9c7","collapsed":true},"outputs":[],"source":["# Look at confusion matrix \n","\n","def plot_confusion_matrix(cm, classes,\n","                          normalize=False,\n","                          title='Confusion matrix',\n","                          cmap=plt.cm.Blues):\n","    \"\"\"\n","    This function prints and plots the confusion matrix.\n","    Normalization can be applied by setting `normalize=True`.\n","    \"\"\"\n","    plt.imshow(cm, interpolation='nearest', cmap=cmap)\n","    plt.title(title)\n","    plt.colorbar()\n","    tick_marks = np.arange(len(classes))\n","    plt.xticks(tick_marks, classes, rotation=45)\n","    plt.yticks(tick_marks, classes)\n","\n","    if normalize:\n","        cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]\n","\n","    thresh = cm.max() / 2.\n","    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):\n","        plt.text(j, i, cm[i, j],\n","                 horizontalalignment=\"center\",\n","                 color=\"white\" if cm[i, j] > thresh else \"black\")\n","\n","    plt.tight_layout()\n","    plt.ylabel('True label')\n","    plt.xlabel('Predicted label')\n","\n","# Predict the values from the validation dataset\n","Y_pred = model.predict(X_val)\n","# Convert predictions classes to one hot vectors \n","Y_pred_classes = np.argmax(Y_pred,axis = 1) \n","# Convert validation observations to one hot vectors\n","Y_true = np.argmax(Y_val,axis = 1) \n","# compute the confusion matrix\n","confusion_mtx = confusion_matrix(Y_true, Y_pred_classes) \n","# plot the confusion matrix\n","plot_confusion_matrix(confusion_mtx, classes = range(10)) "]},{"cell_type":"code","execution_count":null,"metadata":{"_cell_guid":"7b0f31b8-c18b-4529-b0d8-eb4c31e30bbf","_execution_state":"idle","_uuid":"e7a3d6449b499a29db224e42e950f21ca1ec4e36","collapsed":true},"outputs":[],"source":["# Display some error results \n","\n","# Errors are difference between predicted labels and true labels\n","errors = (Y_pred_classes - Y_true != 0)\n","\n","Y_pred_classes_errors = Y_pred_classes[errors]\n","Y_pred_errors = Y_pred[errors]\n","Y_true_errors = Y_true[errors]\n","X_val_errors = X_val[errors]\n","\n","def display_errors(errors_index,img_errors,pred_errors, obs_errors):\n","    \"\"\" This function shows 6 images with their predicted and real labels\"\"\"\n","    n = 0\n","    nrows = 2\n","    ncols = 3\n","    fig, ax = plt.subplots(nrows,ncols,sharex=True,sharey=True)\n","    for row in range(nrows):\n","        for col in range(ncols):\n","            error = errors_index[n]\n","            ax[row,col].imshow((img_errors[error]).reshape((28,28)))\n","            ax[row,col].set_title(\"Predicted label :{}\\nTrue label :{}\".format(pred_errors[error],obs_errors[error]))\n","            n += 1\n","\n","# Probabilities of the wrong predicted numbers\n","Y_pred_errors_prob = np.max(Y_pred_errors,axis = 1)\n","\n","# Predicted probabilities of the true values in the error set\n","true_prob_errors = np.diagonal(np.take(Y_pred_errors, Y_true_errors, axis=1))\n","\n","# Difference between the probability of the predicted label and the true label\n","delta_pred_true_errors = Y_pred_errors_prob - true_prob_errors\n","\n","# Sorted list of the delta prob errors\n","sorted_dela_errors = np.argsort(delta_pred_true_errors)\n","\n","# Top 6 errors \n","most_important_errors = sorted_dela_errors[-6:]\n","\n","# Show the top 6 errors\n","display_errors(most_important_errors, X_val_errors, Y_pred_classes_errors, Y_true_errors)"]},{"cell_type":"code","execution_count":22,"metadata":{"_cell_guid":"05ff3b9f-c3bb-4cec-a8c2-2c128e8f15b3","_execution_state":"idle","_uuid":"7f17e7bf0a54a01a52fef2d554780f6bc6580dc6","collapsed":true},"outputs":[],"source":["# predict results\n","results = model.predict(test)\n","\n","# select the indix with the maximum probability\n","results = np.argmax(results,axis = 1)\n","\n","results = pd.Series(results,name=\"Label\")"]},{"cell_type":"code","execution_count":23,"metadata":{"_cell_guid":"b5f1f39f-13b8-439a-8913-0f120e3d47a9","_execution_state":"idle","_uuid":"369dfaab09240f3f12bcff91953ffd315ab84985","collapsed":true},"outputs":[],"source":["submission = pd.concat([pd.Series(range(1,28001),name = \"ImageId\"),results],axis = 1)\n","\n","submission.to_csv(\"cnn_mnist_datagen.csv\",index=False)"]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.6.1"}},"nbformat":4,"nbformat_minor":2}
