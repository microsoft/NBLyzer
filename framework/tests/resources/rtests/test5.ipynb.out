[{"cell_id":0,"errors":[{"line":1,"label":"", "error_type":"ErrorType.TERMINAL", "message":"Isolated cell."}],"path":[0]},{"cell_id":0,"errors":[{"line":1,"label":"", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[0]},{"cell_id":3,"errors":[{"line":1,"label":"", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[3]},{"cell_id":4,"errors":[{"line":1,"label":"", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[4]},{"cell_id":10,"errors":[{"line":2,"label":"g", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[10]},{"cell_id":13,"errors":[{"line":1,"label":"", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[13]},{"cell_id":18,"errors":[{"line":2,"label":"fig", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":2,"label":"ax", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":5,"label":"legend", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[18]},{"cell_id":20,"errors":[{"line":4,"label":"errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":6,"label":"Y_pred_classes_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":7,"label":"Y_pred_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":8,"label":"Y_true_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":9,"label":"X_val_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":3,"label":"n", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":14,"label":"nrows", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":15,"label":"ncols", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":16,"label":"fig", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":16,"label":"ax", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":1,"label":"error", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":25,"label":"Y_pred_errors_prob", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":28,"label":"true_prob_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":31,"label":"delta_pred_true_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":34,"label":"sorted_dela_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."},{"line":37,"label":"most_important_errors", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[20]},{"cell_id":22,"errors":[{"line":1,"label":"submission", "error_type":"ErrorType.TERMINAL", "message":"Idle cell."}],"path":[22]}]